pyriemann/datasets/sampling.py

```diff
@@ -0,0 +1,338 @@
+        Ui = samples_U[i]
+        ri = samples_r[i]
+        samples[i] = Ui.T @ np.diag(np.exp(ri)) @ Ui
+        samples[i] = (samples[i] + samples[i].T) / 2.0  # ensure symmetry
+
+    return samples
+
+
+def sample_gaussian_spd(n_samples, mean, sigma, random_state=None):
+    """Sample a Riemannian Gaussian distribution
+
+    Sample SPD matrices from a Riemannian Gaussian distribution centered at
+    mean and with dispersion parametrized by sigma. This distribution has been
+    defined in [1]_ and generalizes the notion of a Gaussian distribution to
+    the space of SPD matrices. The sampling is based on a spectral
+    factorization of SPD matrices in terms of their eigenvectors (U-parameters)
+    and the log of the eigenvalues (r-parameters).
+
+    Parameters
+    ----------
+    n_samples : int
+        How many samples to generate.
+    mean : ndarray, shape (n_dim, n_dim)
+        Center of the Riemannian Gaussian distribution.
+    sigma : float
+        Dispersion of the Riemannian Gaussian distribution.
+    random_state : int, RandomState instance or None (default: None)
+        Pass an int for reproducible output across multiple function calls.
+
+    Returns
+    -------
+    samples : ndarray, shape (n_samples, n_dim, n_dim)
+        Samples of the Riemannian Gaussian distribution
+
+    Notes
+    -----
+    .. versionadded:: 0.2.8
+
+    References
+    ----------
+    .. [1] S. Said, L. Bombrun, Y. Berthoumieu, and J. Manton, “Riemannian
+        Gaussian distributions on the space of symmetric positive definite
+        matrices”, IEEE Trans Inf Theory, vol. 63, pp. 2153–2170, 2017.
+        https://arxiv.org/pdf/1507.01760.pdf
+    """
+
+    n_dim = mean.shape[0]
+    samples_centered = sample_gaussian_spd_centered(n_samples,
+                                                    n_dim=n_dim,
+                                                    sigma=sigma,
+                                                    random_state=random_state)
+
+    # apply the parallel transport to mean on each of the samples
+    samples = np.zeros((n_samples, n_dim, n_dim))
+    mean_sqrt = sqrtm(mean)
+    for i in range(n_samples):
+        samples[i] = mean_sqrt @ samples_centered[i] @ mean_sqrt
+        samples[i] = (samples[i] + samples[i].T) / 2.0  # ensure symmetry
+
+    return samples
+
+
+def generate_random_spd_matrix(n_dim, random_state=None):
+    """Generate a random SPD matrix
+
+    Parameters
+    ----------
+    n_dim : int
+        Dimensionality of the matrix to sample.
+    random_state : int, RandomState instance or None (default: None)
+        Pass an int for reproducible output across multiple function calls.
+
+    Returns
+    -------
+    C : ndarray, shape (n_dim, n_dim)
+        Random SPD matrix
+
+    Notes
+    -----
+    .. versionadded:: 0.2.8
+    """
+    rs = check_random_state(random_state)
+    A = rs.randn(n_dim, n_dim)
+    A = (A + A.T)/2
+    _, Q = np.linalg.eig(A)
+    w = rs.rand(n_dim)
```

Comment:
What about using a triangular distribution to generate eigenvalues? @sylvchev

Reply:
@qbarthelemy : Why is that? I am not aware of any particular reason for which a triangular distribution would be more appropriate. Would you have any references I could check on this topic?

Reply:
In real signals, eigenvalues are not uniformly distributed. Triangle distribution allows to avoid too frequent extreme eigenvalues.

But, I have a new idea: since `A` is symmetric, we could simply compute `C = expm(A)`: with this operation, we do not explicitly impose a distribution to eigenvalues.

Reply:
This comes from a method that we used with Quentin for generating SPD matrices, first by creating a reference matrix and adding a perturbation to this reference to generate samples in parametrized neighborhood. The reference matrix is defined to have eigenvalues sampled from triangular distribution, to have few small eigenvalues and possibly large ones.

Reply:
Oooh ok, I see. That's nice indeed, had never thought about it. Thanks for the explanation !

I think I will stick with @qbarthelemy simple solution of using `expm`, we don't really need anything fancier than that for this rather auxiliary function, no?

Reply:
Your new commit is ok for me!